### ISSUE TITLE
ModuleNotFoundError: No module named '_ssl' on Apple Sillicon: "symbol not found in flat namespace '_GENERAL_NAME_free'"

### ISSUE THREAD

**Issue by klueless-io** [pyenv/pyenv#2706]
I have tried multiple times to get Python with PyEnv working on my Mac M2 and always have trouble `openssl`.

I have posted on [Stack Overflow](https://stackoverflow.com/questions/76361324/install-python-3-10-or-3-11-with-openssl-on-mac-apple-silicon)

## Environment

```bash
sw_vers -productName
# macOs

sw_vers -productVersion
# 13.3.1

uname -m
arm64

pyenv --version
# pyenv 2.3.18
```
Before starting, I cleaned out my entire python installations.

## remove any remnants of python

```bash
sudo rm -rf /Library/Frameworks/Python.framework/Versions/3.10
sudo rm -rf /usr/local/bin/python*
sudo rm -rf /usr/local/bin/pip*
sudo rm -rf /usr/local/lib/python*
sudo rm -rf /usr/local/include/python*

sudo rm -rf /Library/Frameworks/Python.framework
sudo rm -rf ~/Library/Caches/pip
sudo rm -rf /Users/<username>/.local/lib/python*
sudo rm -rf /Users/<username>/.cache/pip

brew uninstall python
# done
which python3
# => /usr/bin/python3

python --version
# => Python 3.9.6
```

## Install Python

```bash
brew install python@3.10
brew install python@3.11

python3 --version
Python 3.11.3
```

## Setup pyenv

```bash
brew install pyenv

echo 'if command -v pyenv 1>/dev/null 2>&1; then eval "$(pyenv init -)"; fi' >> ~/.zshrc

source ~/.zshrc
```

## Find out what version of 3.10 and 3.11 to install

```bash
pyenv install --list | grep 3.10
pyenv install --list | grep 3.11

# There is a 3.10.11 and 3.11.3
```

## Use pyenv to install latest versions

> This is where the first issues with SSL start happening

I get exactly the same issue with `3.10.11` and `3.11.3`

```bash
pyenv install 3.11.3
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Installing Python-3.11.3...
python-build: use tcl-tk from homebrew
python-build: use readline from homebrew
python-build: use zlib from xcode sdk
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/xxxxxxxxxx/.pyenv/versions/3.11.3/lib/python3.11/ssl.py", line 100, in <module>
    import _ssl             # if we can't import it, let the error propagate
    ^^^^^^^^^^^
ModuleNotFoundError: No module named '_ssl'
ERROR: The Python ssl extension was not compiled. Missing the OpenSSL lib?

Please consult to the Wiki page to fix the problem.
https://github.com/pyenv/pyenv/wiki/Common-build-problems


BUILD FAILED (OS X 13.3.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/98/545gr7v16mscl5dhgtymzw5r0000gn/T/python-build.20230530123200.52419
Results logged to /var/folders/98/545gr7v16mscl5dhgtymzw5r0000gn/T/python-build.20230530123200.52419.log

Last 10 log lines:
  File "/private/var/folders/98/545gr7v16mscl5dhgtymzw5r0000gn/T/python-build.20230530123200.52419/Python-3.11.3/Lib/hashlib.py", line 123, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type blake2s
Looking in links: /var/folders/98/545gr7v16mscl5dhgtymzw5r0000gn/T/tmpij1ox7r9
Processing /private/var/folders/98/545gr7v16mscl5dhgtymzw5r0000gn/T/tmpij1ox7r9/setuptools-65.5.0-py3-none-any.whl
Processing /private/var/folders/98/545gr7v16mscl5dhgtymzw5r0000gn/T/tmpij1ox7r9/pip-22.3.1-py3-none-any.whl
Installing collected packages: setuptools, pip
  WARNING: The scripts pip3 and pip3.11 are installed in '/Users/xxxxxxxxxx/.pyenv/versions/3.11.3/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed pip-22.3.1 setuptools-65.5.0

```

## Check what version of openssl is installed

```bash
brew list --versions openssl

openssl@3 3.1.0
```

## Testing with `openssl@1.1` and `openssl@3`

```bash
# Attempt #1
ENV | grep FLAGS
LDFLAGS=-L/usr/local/opt/openssl@1.1/lib
CPPFLAGS=-I/usr/local/opt/openssl@1.1/include

# Attempt #2
ENV | grep FLAGS
LDFLAGS=-L/usr/local/opt/openssl@3/lib
CPPFLAGS=-I/usr/local/opt/openssl@3/include
```

## Using openssl@1.1

[![enter image description here][1]][1]

## Using openssl@3

[![enter image description here][2]][2]

## Extra Things I tried after reading your common problems and closed tickets.

I tried removing openssl@3, but apparently pyenv needs it.

```
brew uninstall openssl@3
Error: Refusing to uninstall /opt/homebrew/Cellar/openssl@3/3.1.0
because it is required by pyenv, which is currently installed.
You can override this and force removal with:
  brew uninstall --ignore-dependencies openssl@3
```

I persevered and removed pyenv, followed by openssl@3

```bash
brew uninstall pyenv

# Uninstalling /opt/homebrew/Cellar/openssl@3/3.1.0... (6,494 files, 28.4MB)
```

Then I reinstalled openssl@1.1

```bash
brew install openssl@1.1

# Warning: openssl@1.1 1.1.1t is already installed and up-to-date.
```

Then I reinstalled pyenv and it just reinstalled openssl@3 anyway.

```bash
 brew install pyenv

# Installing dependencies for pyenv: openssl@3
```

So I re-ran `pyenv install 3.10.11` and here is the log

  [1]: https://i.stack.imgur.com/EDyIK.png
  [2]: https://i.stack.imgur.com/rCkau.png

- [ ] pyenv version:
- [ ] Python version:
- [ ] C Compiler information (e.g. gcc 7.3): 
- [ ] Please attach the debug trace of the failing command as a gist:
  * Run `env PYENV_DEBUG=1 <faulty command> 2>&1 | tee trace.log` and attach `trace.log`. E.g. if you have a problem with installing Python, run `env PYENV_DEBUG=1 pyenv install -v <version> 2>&1 | tee trace.log` (note the `-v` option to `pyenv install`).
- [ ] If you have a problem with installing Python, please also attach `config.log` from the build directory
  * The build directory is reported after the "BUILD FAILED" message and is usually under `/tmp`.
- [ ] If the build succeeds but the problem is still with the build process (e.g. the resulting Python is missing a feature), please attach
  * the debug trace from reinstalling the faulty version with `env PYENV_DEBUG=1 pyenv install -f -k -v <version> 2>&1 | tee trace.log`
  * `config.log` from the build directory. When using `pyenv install` with `-k` as per above, the build directory will be under `$PYENV_ROOT/sources`.
[python-build.20230530132102.52442.log](https://github.com/pyenv/pyenv/files/11595439/python-build.20230530132102.52442.log)

**Comment by native-api** [comment]
Duplicate of https://github.com/pyenv/pyenv/issues/2705

**Comment by native-api** [comment]
The same cause as in https://github.com/pyenv/pyenv/issues/2705 : building for ARM64 but using AMD64 Homebrew.

**Comment by native-api** [comment]
Requiring OpenSSL 3 as a dependency is correct because newer versions of CPython support it.

**Comment by ssbarnea** [comment]
@native-api So is there a way to make pyenv build current python versions such 3.11.5 on M1 mac? So far last one that was building was 3.11.4

**Comment by bowen-upenn** [comment]
Use `brew config`  to make sure that your `macOS` looks something like `14.4.1-arm64` rather than `14.4.1-x86_64`. The latter may happen if you are running a terminal session under a Rosetta 2 emulation, even if you have a Mac with the Apple Silicon. If so, start a terminal session explicitly with the ARM architecture using `arch -arm64 /usr/bin/env zsh`.
